### PR TITLE
Set a default memory resources request for Elasticsearch

### PR DESCRIPTION
**Pull request by sebgl**
We recently disabled any default resources: as a result, it is possible
to deploy eg. 30 Elasticsearch nodes on a 1GB machine. They will all
randomly crash (OOM), which is quite confusing.

This commit enables a default memory resources request of 2Gi, matching
the default ES heap size of 1Gi, if the user does not provide their own
resource requirements.

For cases where the user wants to rely on the k8s cluster configured
LimitRanges, hence not provide resource requirements in the spec, it is
possible to explicitly pass empty resource requirements in the spec:

```
podTemplate:
  spec:
    containers:
    - name: elasticsearch
      resources:
        limits: {}
```

In this situation, the operator will not apply default resources.

Fixes https://github.com/elastic/cloud-on-k8s/issues/1097.

### CODE REVIEW

File: operators/pkg/controller/common/defaults/pod_template.go [hunk 1]
```diff
@@ -287,3 +287,14 @@ func (b *PodTemplateBuilder) WithInitContainers(initContainers ...corev1.Contain
 
 	return b
 }
+
+// WithResources sets up the given default resource requirements if both resources limits and requests
```
Comment (thbkrkr) [inline operators/pkg/controller/common/defaults/pod_template.go]: godoc not sync: WithDefaultResources*

Comment (thbkrkr) [reply operators/pkg/controller/common/defaults/pod_template.go]: No strong opinion but I think I would have call it `WithResources(resources corev1.ResourceRequirements)`.

Comment (sebgl) [reply operators/pkg/controller/common/defaults/pod_template.go]: You're right, it better matches other functions of this file: apply our values, unless already set by the user.

**Comment by barkbay** [comment]
Can we improve how the error is surfaced if a limit range is in place _(e.g. with min request > 2G)_ ?
imho the user experience is not good:
1. Apply the CRD
2. Pods are not created ?
3. There are no Kubernetes events related to this cluster
4. There is a stacktrace in the operator log but the name of the cluster is in the `request` field

```
{"level":"error","ts":1562240791.660179,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"elasticsearch-controller","request":"default/quickstart","error":"pods \"quickstart-es-6wq8l9dxt9\" is forbidden: minimum memory usage per Container is 3Gi, but request is 2Gi","errorCauses":[{"error":"pods \"quickstart-es-6wq8l9dxt9\" is forbidden: minimum memory usage per Container is 3Gi, but request is 2Gi"}],"stacktrace":"github.com/elastic/cloud-on-k8s/operators/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/operators/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/elastic/cloud-on-k8s/operators/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/operators/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/elastic/cloud-on-k8s/operators/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/operators/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/elastic/cloud-on-k8s/operators/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/operators/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/elastic/cloud-on-k8s/operators/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/operators/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/elastic/cloud-on-k8s/operators/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/michael/go/src/github.com/elastic/cloud-on-k8s/operators/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
```

We should also add documentation for the workaround.

**Comment by sebgl** [comment]
@barkbay should we emit an event when we fail creating a pod && log our own error (regardless of what the actual error is)?

> We should also add some documentation for the workaround
Definitely. To be covered as part of https://github.com/elastic/cloud-on-k8s/issues/1031.

**Pull request merged**